File: week1/pset1/credit/credit.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>

void checkCard(long number);

int main(void)
{
    // Prompt user for card number
    long number;
    do
    {
        number = get_long("Number: ");
    }
    while (number < 0);

    checkCard(number);

    // long test[16] = {
    //     378282246310005,
    //     371449635398431,
    //     378734493671000,
    //     30569309025904,
    //     6011111111111117,
    //     6011000990139424,
    //     3530111333300000,
    //     3566002020360505,
    //     2221000000000009,
    //     2223000048400011,
    //     2223016768739313,
    //     5555555555554444,
    //     5105105105105100,
    //     4111111111111111,
    //     4012888888881881,
    //     4222222222222
    // };

    // for (int i = 0; i < 16; i++)
    // {
    //     checkCard(test[i]);
    // }
}

// Check card number using Luhn's Algorithm
void checkCard(long number)
{
    // Store number in new variable
    long num = number;

    // Initialize algorithm sum
    int sum = 0;

    // Initialize variable for card number length
    int len = 0;

    // Loop through digits
    while (num > 0)
    {
        // Get digit (from right)
        int digit = num % 10;

        if (len % 2 == 0)
        {
            // If digit position even add digit value to sum
            sum += digit;
        }
        else
        {
            // if digit position odd double and add result digits
            digit *= 2;
            sum += digit / 10 + digit % 10;
        }

        num /= 10;
        len++;
    }

    // Check algorithm sum and number length is valid
    if (sum % 10 == 0 && (len == 13 || len == 15 || len == 16))
    {
        // Get first 2 digits of number
        int start = number / pow(10, len - 2);

        // Check card type
        if (len == 13 || len == 16)
        {
            if (start / 10 == 4)
            {
                printf("VISA\n");
            }
            else if (start >= 51 && start <= 55)
            {
                printf("MASTERCARD\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else if (len == 15)
        {
            if (start == 34 || start == 37)
            {
                printf("AMEX\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

File: week6/pset6/credit/credit.py
# Port of credit.c
import re
from cs50 import get_string


def main():
    # Prompt user for card number
    number = get_string("Number: ")

    # Validate user input (using) str.isdecimal()
    # while not number.isdecimal():
    #     number = get_string("Number: ")

    # Validate user input (using regex)
    while not re.match("^[0-9]+$", number):
        number = get_string("Number: ")

    # Check if number is valid
    check = check_number(number)

    # If valid determine card issuer
    if check:
        if number[0:2] in ["34", "37"]:
            print("AMEX")
        elif number[0] == "4":
            print("VISA")
        elif 51 <= int(number[0:2]) <= 55:
            print("MASTERCARD")
        else:
            print("INVALID")
    else:
        print("INVALID")


def check_number(num):
    """Perform Luhn algorithm on card number."""

    # Get number length and initialize total
    n = len(num)
    total = 0

    # Loop over every other digits from second-to-last
    for i in range(n - 2, -1, -2):
        digit = int(num[i])

        if digit * 2 < 10:
            total += digit * 2
        else:
            total += (digit * 2) % 10
            total += 1

    # Add remaining digits
    for i in range(n - 1, -1, -2):
        digit = int(num[i])
        total += digit

    # If last digit of total is 0 valid else invalid
    return True if total % 10 == 0 else False


if __name__ == "__main__":
    main()